### PR TITLE
 port remaining parameter definition widgets to C++

### PR DESCRIPTION
**Pull request by alexbruy**
## Description
Port remaining modeler parameter definition widgets to C++. This include:

- band (with exposed option for multiple bands)
- field (with exposed option for selecting all fields by default)
- multiple layers
- vector layer
- feature source
- number
- scale
- distance
- range (with exposed range boundaries)
- map layer
 
For map layer, vector layer and feature source parameters enable multiple layer types support.

### CODE REVIEW

File: src/gui/qgscheckablecombobox.cpp [hunk 1]
```diff
@@ -148,6 +148,24 @@ QStringList QgsCheckableComboBox::checkedItems() const
   return items;
 }
 
+QVariantList QgsCheckableComboBox::checkedItemsData() const
```
Comment (Gustry) [inline src/gui/qgscheckablecombobox.cpp:151]: Hi, did you see this ticket? https://github.com/qgis/QGIS/issues/28658

File: src/gui/processing/qgsprocessingwidgetwrapperimpl.cpp [hunk 2]
```diff
@@ -613,6 +614,65 @@ QgsAbstractProcessingParameterWidgetWrapper *QgsProcessingAuthConfigWidgetWrappe
...
+    mDefaultLineEdit->setText( numberParam->defaultValue().toString() );
+  }
+
+  setLayout( vlayout );
+}
+
+QgsProcessingParameterDefinition *QgsProcessingNumberParameterDefinitionWidget::createParameter( const QString &name, const QString &description, QgsProcessingParameterDefinition::Flags flags ) const
+{
+  QgsProcessingParameterNumber::Type dataType = static_cast< QgsProcessingParameterNumber::Type >( mTypeComboBox->currentData().toInt() );
+  auto param = qgis::make_unique< QgsProcessingParameterNumber >( name, description, dataType, mDefaultLineEdit->text() );
```
Comment (nyalldawson) [inline src/gui/processing/qgsprocessingwidgetwrapperimpl.cpp:657]: We should convert mDefaultLineEdit->text() to a double if possible

File: src/gui/processing/qgsprocessingwidgetwrapperimpl.cpp [hunk 3]
```diff
@@ -613,6 +614,65 @@ QgsAbstractProcessingParameterWidgetWrapper *QgsProcessingAuthConfigWidgetWrappe
...
+
+  bool ok;
+  float val = mMinLineEdit->text().toFloat( &ok );
```
Comment (nyalldawson) [inline src/gui/processing/qgsprocessingwidgetwrapperimpl.cpp:660]: float -> double

File: src/gui/processing/qgsprocessingwidgetwrapperimpl.cpp [hunk 4]
```diff
@@ -837,10 +897,123 @@ QgsAbstractProcessingParameterWidgetWrapper *QgsProcessingNumericWidgetWrapper::
...
+    mMaxLineEdit->setText( QString::number( distParam->maximum() ) );
+    mDefaultLineEdit->setText( distParam->defaultValue().toString() );
+  }
+
+  setLayout( vlayout );
+}
+
+QgsProcessingParameterDefinition *QgsProcessingDistanceParameterDefinitionWidget::createParameter( const QString &name, const QString &description, QgsProcessingParameterDefinition::Flags flags ) const
+{
+  auto param = qgis::make_unique< QgsProcessingParameterDistance >( name, description, mDefaultLineEdit->text(), mParentLayerComboBox->currentData().toString() );
```
Comment (nyalldawson) [inline src/gui/processing/qgsprocessingwidgetwrapperimpl.cpp:998]: same notes as for the numeric parameter

File: src/gui/processing/qgsprocessingwidgetwrapperimpl.cpp [hunk 5]
```diff
@@ -3402,6 +3678,105 @@ void QgsProcessingFieldPanelWidget::updateSummaryText()
...
+  vlayout->addWidget( mParentLayerComboBox );
+
+  vlayout->addWidget( new QLabel( tr( "Allowed data type" ) ) );
+  mDataTypeComboBox = new QComboBox();
+  mDataTypeComboBox->addItem( tr( "Any" ), QgsProcessingParameterField::Any );
+  mDataTypeComboBox->addItem( tr( "Number" ), QgsProcessingParameterField::Numeric );
+  mDataTypeComboBox->addItem( tr( "String" ), QgsProcessingParameterField::String );
+  mDataTypeComboBox->addItem( tr( "Date/time" ), QgsProcessingParameterField::DateTime );
+  if ( const QgsProcessingParameterField *fieldParam = dynamic_cast<const QgsProcessingParameterField *>( definition ) )
+    mDataTypeComboBox->setCurrentIndex( fieldParam->dataType() + 1 ); // QComboBoxes indexes start at 0, datatype start with -1 that is why I need to do +1
```
Comment (nyalldawson) [inline src/gui/processing/qgsprocessingwidgetwrapperimpl.cpp:3736]: would be nicer to use findData here

**Review by nyalldawson — COMMENTED** [review]
Really nice! This will allow for some more much needed cleanups, as we can drop the whole ModelerParameterDefinitionDialog python class...

File: src/gui/processing/qgsprocessingwidgetwrapperimpl.cpp (hunk 5 above)
Comment (alexbruy) [reply src/gui/processing/qgsprocessingwidgetwrapperimpl.cpp]: Done

File: src/gui/processing/qgsprocessingwidgetwrapperimpl.cpp (hunk 4 above)
Comment (alexbruy) [reply src/gui/processing/qgsprocessingwidgetwrapperimpl.cpp]: Fixed

File: src/gui/processing/qgsprocessingwidgetwrapperimpl.cpp (hunk 3 above)
Comment (alexbruy) [reply src/gui/processing/qgsprocessingwidgetwrapperimpl.cpp]: Fixed

File: src/gui/processing/qgsprocessingwidgetwrapperimpl.cpp (hunk 2 above)
Comment (alexbruy) [reply src/gui/processing/qgsprocessingwidgetwrapperimpl.cpp]: Fixed

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**